ql/src/java/org/apache/hadoop/hive/ql/queryhistory/QueryHistoryService.java

```diff
@@ -0,0 +1,238 @@
+      throw new RuntimeException("QueryHistoryService is not present when asked for the singleton instance");
+    }
+    return INSTANCE;
+  }
+
+  private void printConfigInformation() {
+    if (batchSize == 0) {
+      LOG.info("Query History batch size is 0: every record will be flushed synchronously after the query" +
+          " (not recommended)");
+    } else {
+      LOG.info("Query History batch size is {}: records will be flushed after every {} queries",
+          batchSize, batchSize);
+    }
+    if (maxQueueSizeInMemoryBytes == 0) {
+      LOG.info("Query History max queue memory byte size is 0: the service won't check the consumed memory to make " +
+          "flush decision (not recommended)");
+    } else {
+      LOG.info("Query History max queue memory byte size is {}: records will be flushed when the queue consumes " +
+              "more than {} bytes in the memory",
+          maxQueueSizeInMemoryBytes, maxQueueSizeInMemoryBytes);
+    }
+
+    if (flushInterval == 0) {
+      LOG.info("Query History flush interval size is 0: the service won't check for records to flush periodically");
+    } else {
+      LOG.info("Query History flush interval is {}s: service will attempt to flush records in every {}s",
+          flushInterval, flushInterval);
+    }
+  }
+
+  private QueryHistoryRepository createRepository(HiveConf conf) {
+    try {
+      return (QueryHistoryRepository) ReflectionUtil.newInstance(
+          conf.getClassByName(conf.get(HiveConf.ConfVars.HIVE_QUERY_HISTORY_REPOSITORY_CLASS.varname)), conf);
+    } catch (ClassNotFoundException e) {
+      throw new RuntimeException(e);
+    }
+  }
+
+  /**
+   * Creates a Query History record from a given DriverContext and adds it to the queue.
+   * This is done asynchronously to completely decouple from the threads that run the query to make this service work
+   * silently in the background.
+   * @param driverContext a DriverContext containing all the needed information for a record to be created
+   */
+  public void logQuery(DriverContext driverContext) {
+    long recordCreateStart = Time.monotonicNow();
+    Record record = createRecord(driverContext);
+    long recordCreateEnd = Time.monotonicNow();
+    queryHistoryQueue.add(record);
+    LOG.info("Created query history record in {}ms, current queue size: {}",
+        recordCreateEnd - recordCreateStart, queryHistoryQueue.size());
+
+    CompletableFuture<?> flushFuture = flushRecordsIfNeeded();
+    maybeWaitForRecordsToBeFlushed(batchSize == 0, flushFuture);
+  }
+
+  private CompletableFuture<?> flushRecordsIfNeeded() {
+    if (needToFlush()) {
+      return doFlush();
+    } else {
+      LOG.debug("Not flushing history records yet, current queue size: {}, batchSize: {}",
+          queryHistoryQueue.size(), batchSize);
+      return CompletableFuture.completedFuture(null);
+    }
+  }
+
+  private void maybeWaitForRecordsToBeFlushed(boolean shouldWait, CompletableFuture<?> flushFuture) {
```

Comment:
can we drop `maybe` from the name, just `wait`

Reply:
you mean waitForRecordsToBeFlushed? the rest of the name is ok?

Reply:
yes